### DISCUSSION TITLE
Runtime descriptor parsing from hidapi

### DISCUSSION THREAD

**Discussion by peter-vbet** [IntergatedCircuits/hid-rp#9] [Q&A] [answered]
Hi, can I use this library to parse the report descriptor acquired from this hidapi librabry for Headset device https://github.com/libusb/hidapi/pull/451/files#diff-37a51f8fe6a77e89ab9694a39cf4e6c665cbdc96203d35ead0f55a9c37e0fc25 , is there any documentation for this library. thanks

**Comment by benedekkupper** [comment]
Yes you can!

The [`parser`](https://github.com/IntergatedCircuits/hid-rp/blob/master/hid-rp/hid/rdf/parser.hpp) class is a generic template for achieving any kind of HID report descriptor parsing. It provides the necessary encapsulation while leaving the specifics to its subclasses. So far - to my knowledge - this library has only been used for descriptor parsing at compile-time, helping the HID device FW implementation, so there is no prior work done that illustrates a full example of descriptor parsing at runtime. The exception being the [`get_application_usage_id()`](https://github.com/IntergatedCircuits/hid-rp/blob/master/hid-rp/hid/rdf/parser.hpp#L348) method, this is something you can already use, for the purpose of finding the HID headset device from the list of HID devices available to the system.

Another reference for your implementation is [`report_protocol_properties::parser`](https://github.com/IntergatedCircuits/hid-rp/blob/dc2c953e6d5914d9cbc5ab621aff93d7e06dfd61/hid-rp/hid/report_protocol.hpp#L122) class, which while intended for compile-time use, is still a good example of how to extract important information from the HID report descriptor. The key difference is that compile-time descriptor parsing cannot rely on dynamic memory allocation, while for runtime parsing that's the highly advised way to go.

If you provide more specific details on what information you want to gain out of the descriptor, I can guide you further.

**Comment by peter-vbet** [comment]
Hi, I want to parse the descriptor at run-time on desktop end to a readable format in c++ instad of raw bytes, for example like the javascript webhid object strucuture implemented by google https://wicg.github.io/webhid/#hidreportitem-dictionary, so that I clearly read and do some computation based on some fields

**Comment by benedekkupper** [comment, marked as answer]
For that the best starting point is here:
https://github.com/IntergatedCircuits/hid-rp/blob/dc2c953e6d5914d9cbc5ab621aff93d7e06dfd61/hid-rp/hid/report_protocol.hpp#L180-L234
You need to replace the internal logic to push a new item to a vector of HID report items, filled with the information you get out of the passed parameters. The vector is a member variable of your custom parser, that's how it can be accessed reasonably in this virtual function. You get the boolean values out of the `main_item` parameter, the usages are collected by iterating through the `main_section` (see also the aforementioned `get_application_usage_id`), the rest are acquired from `global_state`.
